File: BullCowGame/Swift/BullCowGame.swift
import Foundation

struct BullCowCount {
    var bulls = 0
    var cows = 0
}

enum GuessStatus {
    case Invalid
    case Ok
    case NotIsogram
    case WrongLength
    case NotLowercase
}

struct BullCowGame {
    private var currentTry: Int = 0
    private var gameIsWon: Bool = false
    private var hiddenWord: String = ""
    
    init() {
        self.reset()
    }
    
    mutating func reset() {
        currentTry = 1
        hiddenWord = self.fetchHiddenWord()
        gameIsWon = false
    }
    
    private func fetchHiddenWord() -> String {
        let file: FileHandle? = FileHandle(forReadingAtPath: "words.txt")
        var res: String?
        if file != nil {
            let data = file!.readDataToEndOfFile()
            file!.closeFile()
            let str = NSString(data: data, encoding: 4) as String?
            let arr = str!.components(separatedBy:"\n")
            let rand = Int(arc4random_uniform(UInt32(arr.count - 1)))
            res = arr[rand]
        }
        return res ?? ""
    }
    
    private func isIsogram(_ guess: String) -> Bool {
        if guess.count <= 1 {
            return true
        }
        var letterToHaveBeenSeen = [Character: Bool]()
        let temp = guess.lowercased()
        for letter in temp {
            if letterToHaveBeenSeen[letter] ?? false {
                return false
            } else {
                letterToHaveBeenSeen[letter] = true
            }
        }
        return true
    }
    
    private func isLowercase(_ guess: String) -> Bool {
        var isLowercased: Bool = false
        for letter in guess {
            isLowercased = letter.isLowercase == true
        }
        return isLowercased
    }
    
    func getMaxTries() -> Int {
        let wordLengthToMaxTries: [Int: Int] = [3:4, 4:7, 5:10, 6:16, 7:20]
        return wordLengthToMaxTries[hiddenWord.count]!
    }
    
    func getCurrentTry() -> Int {
        return currentTry
    }
    
    func getHiddenWordLength() -> Int {
        return hiddenWord.count
    }
    
    func isGameWon() -> Bool {
        return gameIsWon
    }
    
    func checkGuessValidity(_ guess: String) -> GuessStatus {
        if guess.count != self.getHiddenWordLength() {
            return .WrongLength
        } else if !self.isIsogram(guess) {
            return .NotIsogram
        } else if !self.isLowercase(guess) {
            return .NotLowercase
        }
        return .Ok
    }
    
    mutating func submitValidGuess(_ guess: String) -> BullCowCount {
        currentTry = currentTry + 1
        var bullCowCount = BullCowCount()
        for i in 0..<self.getHiddenWordLength() {
            for j in 0..<self.getHiddenWordLength() {
                if guess[guess.index(guess.startIndex, offsetBy: j)] ==
                    hiddenWord[hiddenWord.index(hiddenWord.startIndex, offsetBy: i)] {
                    if i == j {
                        bullCowCount.bulls = bullCowCount.bulls + 1
                    } else {
                        bullCowCount.cows = bullCowCount.cows + 1
                    }
                }
            }
        }
        gameIsWon = bullCowCount.bulls == self.getHiddenWordLength()
        return bullCowCount
    }
}

File: BullCowGame/C++/BullCowGame.cc
#include "BullCowGame.hh"
#include <fstream>
#include <ctime>
#include <map>
#include <vector>
#include <string>

using namespace std;

BullCowGame::BullCowGame() {
    reset();
}

bool BullCowGame::isIsogram(const string guess) const {
    if (guess.length() <= 1)
        return true;
    map<char, bool> letterToHaveBeenSeen;
    for (auto letter: guess) {
        letter = tolower(letter);
        if (letterToHaveBeenSeen[letter])
            return false;
        letterToHaveBeenSeen[letter] = true;
    }
    return true;
}

bool BullCowGame::isLowercase(const string guess) const {
    bool isLowCase;
    for (auto letter: guess)
        isLowCase = islower(letter) == true;
    return isLowCase;
}

bool BullCowGame::isGameWon() const {
    return gameIsWon;
}

int BullCowGame::getCurrentTry() const {
    return currentTry;
}

int BullCowGame::getHiddenWordLength() const {
    return hiddenWord.length();
}

int BullCowGame::getMaxTries() const {
    map<int, int> wordLengthToMaxTries = {{3,4}, {4,7}, {5,10}, {6,16}, {7,20}};
    return wordLengthToMaxTries[hiddenWord.length()];
}

string BullCowGame::fetchHiddenWord() const {
    vector<string> words;
    string word = "";
    char ch;
    ifstream fin("words.txt");
    while (fin.get(ch) && !fin.eof()) {
        if (ch == '\n') {
            words.push_back(word);
            word = "";
            continue;
        }
        word += ch;
    }
    fin.close();
    srand(time(NULL));
    return words[rand() % words.size()];
}

void BullCowGame::reset() {
    currentTry = 1;
    hiddenWord = fetchHiddenWord();
    gameIsWon = false;
}

GuessStatus BullCowGame::checkGuessValidity(const string guess) const {
    if (guess.length() != getHiddenWordLength())
        return GuessStatus::WRONG_LENGTH;
    else if (!isIsogram(guess))
        return GuessStatus::NOT_ISOGRAM;
    else if (!isLowercase(guess))
        return GuessStatus::NOT_LOWERCASE;
    return GuessStatus::OK;
}

BullCowCount BullCowGame::submitValidGuess(const string guess) {
    currentTry++;
    BullCowCount bullCowCount;
    const int WORD_LENGTH = hiddenWord.length();
    for (int i = 0; i < WORD_LENGTH; i++)
        for (int j = 0; j < WORD_LENGTH; j++)
            if (guess[j] == hiddenWord[i]) {
                if (i == j)
                    bullCowCount.bulls++;
                else
                    bullCowCount.cows++;
            }
    gameIsWon = bullCowCount.bulls == WORD_LENGTH;
    return bullCowCount;
}
